src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/WebTransportTests.cs

```diff
@@ -0,0 +1,129 @@
+using Http3SettingType = Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3.Http3SettingType;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests;
+
+public class WebTransportTests : Http3TestBase
+{
+    [Fact]
+    public async Task WebTransportHandshake_ClientToServerPasses()
+    {
+        await Http3Api.InitializeConnectionAsync(_noopApplication);
+        var controlStream = await Http3Api.CreateControlStream();
+        var controlStream2 = await Http3Api.GetInboundControlStream();
+
+        var settings = new Http3PeerSettings()
+        {
+            EnableWebTransport = 1,
+            H3Datagram = 1,
+        };
+
+        await controlStream.SendSettingsAsync(settings.GetNonProtocolDefaults());
+        var response1 = await controlStream2.ExpectSettingsAsync();
+
+        await Http3Api.ServerReceivedSettingsReader.ReadAsync().DefaultTimeout();
+
+        Assert.Equal(1, response1[(long)Http3SettingType.EnableWebTransport]);
+
+        var requestStream = await Http3Api.CreateRequestStream();
+        var headersConnectFrame = new[]
+        {
+            new KeyValuePair<string, string>(HeaderNames.Method, "CONNECT"),
+            new KeyValuePair<string, string>(HeaderNames.Protocol, "webtransport"),
+            new KeyValuePair<string, string>(HeaderNames.Scheme, "http"),
+            new KeyValuePair<string, string>(HeaderNames.Path, "/"),
+            new KeyValuePair<string, string>(HeaderNames.Authority, "server.example.com"),
+            new KeyValuePair<string, string>(HeaderNames.Origin, "server.example.com")
+        };
+
+        await requestStream.SendHeadersAsync(headersConnectFrame);
+        var response2 = await requestStream.ExpectHeadersAsync();
+
+        Assert.Equal((int)HttpStatusCode.OK, Convert.ToInt32(response2[HeaderNames.Status], null));
+
+        await requestStream.OnDisposedTask.DefaultTimeout();
+    }
+
+    [Theory]
+    [InlineData(
+        ((long)Http3ErrorCode.ProtocolError),
+        nameof(HeaderNames.Method), "GET", // incorrect method (verifies that webtransport doesn't break regular Http/3 get)
+        nameof(HeaderNames.Protocol), "webtransport",
+        nameof(HeaderNames.Scheme), "http",
+        nameof(HeaderNames.Path), "/",
+        nameof(HeaderNames.Authority), "server.example.com",
+        nameof(HeaderNames.Origin), "server.example.com")]
+    //[InlineData(
+    //    ((long)Http3ErrorCode..ProtocolError,
+    //    nameof(HeaderNames.Method), "CONNECT",
+    //    nameof(HeaderNames.Protocol), "webtransport",
+    //    nameof(HeaderNames.Scheme), "http", // incorrect scheme <-- it should be https but the tests don't work with it so I commented it out (also it is given by http/3)
+    //    nameof(HeaderNames.Path), "/",
+    //    nameof(HeaderNames.Authority), "server.example.com",
+    //    nameof(HeaderNames.Origin), "server.example.com")]
+    [InlineData(
+        ((long)Http3ErrorCode.ProtocolError),
+        nameof(HeaderNames.Method), "CONNECT",
```

Comment:
```suggestion
        HeaderNames.Method, "CONNECT",
```

This would let you delete GetHeaderFromName too